src/domains/TypesDomain.js

```diff
@@ -50,8 +51,14 @@ export default class TypesDomain {
         if (lType === undefined || rType === undefined) return TypesDomain.topVal;
         if (Value.isTypeCompatibleWith(lType, StringValue) || Value.isTypeCompatibleWith(rType, StringValue))
           resultType = StringValue;
-        else if (Value.isTypeCompatibleWith(lType, NumberValue) && Value.isTypeCompatibleWith(rType, NumberValue))
-          resultType = NumberValue;
+        else if (Value.isTypeCompatibleWith(lType, IntegralValue) && Value.isTypeCompatibleWith(rType, IntegralValue))
+          resultType = IntegralValue;
+        else if (
+          ((Value.isTypeCompatibleWith(lType, NumberValue) || Value.isTypeCompatibleWith(lType, IntegralValue)) &&
```

Comment:
Value.isTypeCompatibleWith will work out that an IntegralValue instance is also a NumberValue instance, so there is no need to check for the former. Alternatively, one could use === instead of calling Value.isTypeCompatibleWith.